src/lib/transition_frontier/persistent_frontier/worker.ml

```diff
@@ -124,21 +116,58 @@ module Worker = struct
         deferred_result_list_fold t ~init ~f
 
   let perform t input =
+    let arcs_cache = State_hash.Table.create () in
     O1trace.thread "persistent_frontier_write_to_disk" (fun () ->
-        match%map
-          [%log' trace t.logger] "Applying %d diffs to the persistent frontier"
-            (List.length input) ;
-          (* Iterating over the diff application in this way
-             * effectively allows the scheduler to scheduler
-             * other tasks in between diff applications.
-             * If implemented otherwise, all diffs would be
-             * applied during the same scheduler cycle.
-          *)
-          deferred_result_list_fold input ~init:() ~f:(fun () diff ->
-              Deferred.return (handle_diff t diff) )
-        with
-        | Ok () ->
-            ()
+        [%log' trace t.logger] "Applying %d diffs to the persistent frontier"
+          (List.length input) ;
+        let best_tip_cnt, root_cnt =
+          List.fold input ~init:(0, 0) ~f:(fun (b, r) -> function
+            | Diff.Lite.E.E (Root_transitioned _) ->
+                (b, r + 1)
+            | E (Best_tip_changed _) ->
+                (b + 1, r)
+            | _ ->
+                (b, r) )
+        in
+        let best_tip_cnt = ref best_tip_cnt in
+        let root_cnt = ref root_cnt in
+        let garbage_prev = ref [] in
+        let input =
+          List.filter_map input ~f:(function
```

Comment:
I find this code a bit tricky to reason about, even though I think it's correct. I'd probably prefer this to be written somewhat as follows thought:

- partion diffs by type (best tip changed, root transitioned, others)
- perform the diff optimizations by folding over the all elements of best tip changed except for the last (do the same with root transitioned)

I think the code will be easier to reason about if implemented this way, instead of having to count the diff types and using mutable logic in the loop to reason about which element is the last in the sequence of those types.

Reply:
@nholland94 I implemented it with counting because I didn't want this code to rely on that last two items should be best tip change and root transitioned. And if I partition by type, how could I reconstruct the order later?

Reply:
Not quite. Let me take a stab at what I was describing and see if I can come up with something a bit clearer. Though if it ends up being more of a hassle than I think it will be, we can just defer refactoring this until later.